src/runtime/sync.go

```diff
@@ -1,13 +1,72 @@
+// Copyright 2009 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
 package runtime
 
+import (
+	"sync/atomic"
+	"unsafe"
+)
+
 // This file contains stub implementations for internal/poll.
+// The official golang implementation states:
+//
+// "That is, don't think of these as semaphores.
+// Think of them as a way to implement sleep and wakeup
+// such that every sleep is paired with a single wakeup,
+// even if, due to races, the wakeup happens before the sleep."
```

Comment:
Ah, that's useful to know. So they're not really semaphores but something like it.

Reply:
https://cs.opensource.google/go/go/+/master:src/runtime/sema.go;l=12?q=semac&ss=go%2Fgo